lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp

```diff
@@ -0,0 +1,299 @@
+  SmallVector<Value> preceedingOperands;
+  auto invokeCall = xsmm::utils::buildInvokeCall(
+      rewriter, output, module, inputs, preceedingOperands, -1, operandRange,
+      invokeName, dtype, true);
+  return std::make_pair(&*dispatched, &*invokeCall);
+}
+
+static LogicalResult
+validateTransposeOpImpl(PatternRewriter &rewriter, Operation *transposeOp,
+                        Operation *input, Operation *output, Type outputType) {
+  LLVM_DEBUG(llvm::dbgs() << "validateTransposeOpImpl\n");
+  Value result = input->getResult(0);
+  Value source = input->getOperand(0);
+  VectorType outType = cast<VectorType>(outputType);
+  VectorType resultType = cast<VectorType>(result.getType());
+  if (!outType.hasStaticShape() || !resultType.hasStaticShape()) {
+    return failure();
+  }
+  if (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::TRANSPOSE,
+                                  outType)) {
+    memref::ExpandShapeOp expandShapeOp =
+        dyn_cast<memref::ExpandShapeOp>(source.getDefiningOp());
+    if (!expandShapeOp || expandShapeOp.getSrcType().getRank() != 2) {
+      return failure(transposeOp);
+    }
+    source = expandShapeOp.getSrc();
+    auto stridesOnInput = mlir::utils::getStaticStrides(source);
+    if (failed(stridesOnInput) || stridesOnInput->back() != 1) {
+      return failure(transposeOp);
+    }
+    auto stridesOnOutput = mlir::utils::getStaticStrides(output->getOperand(1));
+    if (failed(stridesOnOutput)) {
+      return failure(transposeOp);
+    }
+    auto unaryInfo = xsmm::utils::getVectorUnaryInfo(
+        dyn_cast<MemRefType>(input->getOperand(0).getType()),
+        dyn_cast<MemRefType>(output->getOperand(1).getType()),
+        dyn_cast<VectorType>(input->getResult(0).getType()),
+        dyn_cast<VectorType>(output->getOperand(0).getType()),
+        xsmm::UnaryFlags::NONE);
+    if (failed(unaryInfo)) {
+      return failure(transposeOp);
+    }
+  } else {
+    if (!xsmm::utils::isTwoDTransposeOp(
+            dyn_cast<mlir::vector::TransposeOp>(transposeOp))) {
+      return failure(transposeOp);
+    }
+    auto unaryInfo = xsmm::utils::getVectorUnaryInfo(
+        dyn_cast<MemRefType>(input->getOperand(0).getType()),
+        dyn_cast<MemRefType>(output->getOperand(1).getType()),
+        dyn_cast<VectorType>(input->getResult(0).getType()),
+        dyn_cast<VectorType>(output->getOperand(0).getType()),
+        xsmm::UnaryFlags::NONE);
+
+    if (failed(unaryInfo)) {
+      return failure(transposeOp);
+    }
+  }
+  return success(transposeOp);
+}
+
+void registerNativeTransposeRewrite(RewritePatternSet &patterns) {
+  patterns.getPDLPatterns().registerRewriteFunction("BuildTranspose",
+                                                    buildTransposeOp);
+  patterns.getPDLPatterns().registerConstraintFunction("ValidateTranspose",
+                                                       validateTransposeOpImpl);
+}
+
+namespace mlir {
+namespace tpp {
+
+struct ConvertTranspose
```

Comment:
Actually, this one is just the transpose conversion, because we want to prioritize transpose conversions over other ops, and we don't know how else to force conversion of transposes first before converting other ops. That's the reason transposes are in a separate pdll and the rest of the ops are in a subsequent pdll.

Reply:
> we want to prioritize transpose conversions over other ops

I'm not sure about this. Can you expand?

> That's the reason transposes are in a separate pdll

I understood we were just adding transpose first because it's easier and allows us to review the infrastructure before the complex ops. I don't see any special reason why the transpose should be before any other.

But even if so, there has to be an order to PDL patterns that is not individually running them. For instance, we want to look at fused GEMMs before plain GEMMs, softmax before a sequence of element-wise and reduction, etc.

Reply:
As discussed, we'll be moving transposes and other vector ops into a single pdll file for now.